Graphs

laf opened the pull request:
Please see our [contributing docs](https://github.com/librenms/librenmsv2/blob/develop/docs/General/Contributing.md)

- [x] Have you created any necessary docs?
- [x] Have you created tests for Travis integration?
- [x] Have you used localisation in the WebUI?
- [x] Have you included the proper copyright?

Won't be the best code just yet but gets the basics down and adds some examples of:

- Adding a datatables page without widgets but on a page with other elements.
- Adding graphs with varying options to produce png or js graphs (only device_bits has both options available).
-  Started the device overview page + it's file structure.

![image](https://cloud.githubusercontent.com/assets/3941142/20248849/6fef2866-a9e4-11e6-9957-94d6b258da15.png)

laf committed:
Initial graph support

laf committed:
More graph updates

laf committed:
Merge branch 'develop' of github.com:librenms/librenmsv2 into graphs

laf committed:
Merge branch 'develop' of github.com:librenms/librenmsv2 into graphs

laf committed:
More graph updates

laf committed:
More updates to graphs

laf committed:
Merge branch 'develop' of github.com:librenms/librenmsv2 into graphs

laf committed:
More updates

laf committed:
Updates

laf committed:
updated

laf committed:
Merge branch 'develop' of github.com:laf/librenmsv2 into graphs

laf committed:
More updates

laf committed:
Merge branch 'develop' of github.com:librenms/librenmsv2 into graphs

laf committed:
More updates

murrant commented:
Missing the file `resources/views/devices/page/graphs.blade.php`

laf commented:
That page isn't done, I just created the tabs for visual reference. Same with 'Health'

murrant commented:
Oh, thanks I'll check again.  I had to wait for more graph data to collect as I didn't have much on my test system...

murrant commented:
Well, I did `composer dump-autoload` and set $config['install_dir'] and it is working :)
I've been hacking on it a bit...

Not sure the helpers approach is the best, but I can understand pushing dealing with those until later.

Jellyfrog commented on app/Helpers/Rewrite.php:
```diff
@@ -0,0 +1,20 @@
+    $bad_descr = array(
+        '/GenuineIntel:/' => '',
+        '/AuthenticAMD:/' => '',
+        '/Intel(R)/' => '',
```
Typo? `Intel\(R\)`

murrant replied:
No, these are regex.  For preg_replace.

Jellyfrog replied:
yes, which would match the string "_IntelR_", but I'm pretty sure @laf is trying to match "_Intel(R)_", else it makes no sense to include the parentheses

murrant replied:
Hmm, this is legacy code, perhaps it has always been broken...

Jellyfrog commented on app/Helpers/Rewrite.php:
```diff
@@ -0,0 +1,20 @@
+        '/AuthenticAMD:/' => '',
+        '/Intel(R)/' => '',
+        '/CPU/' => '',
+        '/(R)/' => '',
```
Typo? `\(R\)`

murrant committed:
Take advantage of object-oriented programming to simplify things more
Minimize method exposure, so only the ones we want are called by outside code
Properties of the parent class are accessible from children
Throw exception when a graph class isn't found
Add some more Exceptions

murrant committed:
Push buildRRDXport and buildRRDGraphParams down into the buildRRDJson and buildRRDGraph functions.
Device location and hostname are public properties, we don't need getter functions (if we need to modify the output, we can use an accessor function)

murrant committed:
Reorganize classes
Implement Data Source classes, move code into those classes.

murrant committed:
Fix error when os.<name>.over is undefined. For example `os.generic.over`,

murrant committed:
Re-organize graphs.
Move class resolver into static function on Graph.php

murrant committed:
Simplify json and csv serialize.
Add labels to json
Fix json timestamps
(int)null == 0

laf commented:
Thanks murrant :)

murrant commented:
Thank you great work.
